src/libraries/System.IO.FileSystem/tests/FileStream/DeleteOnClose.cs

```diff
@@ -0,0 +1,82 @@
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Tests
+{
+    public class FileStream_DeleteOnClose : FileSystemTest
+    {
+        [InlineData(FileMode.Append)] // FileModes that open an existing file, or create a new one when it doesn't exist.
+        [InlineData(FileMode.Create)]
+        [InlineData(FileMode.OpenOrCreate)]
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsFileLockingEnabled), nameof(PlatformDetection.IsThreadingSupported))]
+        public async Task DeleteOnClose_UsableAsMutex(FileMode mode)
+        {
+            var cts = new CancellationTokenSource();
+            int enterCount = 0;
+            int locksRemaining = int.MaxValue;
+            bool exclusive = true;
+
+            string path = GetTestFilePath();
+            Assert.False(File.Exists(path));
+
+            Func<Task> lockFile = async () =>
+            {
+                while (!cts.IsCancellationRequested)
+                {
+                    try
+                    {
+                        using (var fs = new FileStream(path, mode, FileAccess.Write, FileShare.None, 4096, FileOptions.DeleteOnClose))
+                        {
+                            int counter = Interlocked.Increment(ref enterCount);
+                            if (counter != 1)
+                            {
+                                exclusive = false;
+                                cts.Cancel();
+                                return;
+                            }
+
+                            // Hold the lock for a little bit.
+                            await Task.Delay(TimeSpan.FromMilliseconds(5));
+
+                            Interlocked.Decrement(ref enterCount);
+
+                            if (Interlocked.Decrement(ref locksRemaining) <= 0)
+                            {
+                                return;
+                            }
+                        }
+                        await Task.Delay(TimeSpan.FromMilliseconds(1));
+                    }
+                    catch (UnauthorizedAccessException)
+                    {
+                        // This can occur when the file is being deleted on Windows.
+                        await Task.Delay(TimeSpan.FromMilliseconds(1));
+                    }
+                    catch (IOException)
+                    {
+                        await Task.Delay(TimeSpan.FromMilliseconds(1));
+                    }
+                }
+            };
+
+            var tasks = new Task[50];
+            for (int i = 0; i < tasks.Length; i++)
+            {
+                tasks[i] = Task.Run(lockFile);
+            }
+
+            // Wait for 1000 locks.
+            cts.CancelAfter(TimeSpan.FromSeconds(100));
+            Volatile.Write(ref locksRemaining, 500);
+            await Task.WhenAll(tasks);
+
+            Assert.True(exclusive, "Exclusive");
+            Assert.False(cts.IsCancellationRequested, "Test cancelled");
```

Comment:
If I am understanding this correctly, the `cts` gets cancelled if we can't reserve 1000 locks. How often do you think our test machines will fail to reserve all those locks? Could it be a potential cause of intermittent failures?

Reply:
The `cts` gets cancelled when the the locks is not exclusive (test fails). This should never happen.
Or if the test takes longer than a timeout as set by `CancelAfter`. The timeout should be properly chosen so this doesn't happen on the test machines.